mempool/src/shared_mempool/use_case_history.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_logger::info;
+use aptos_mempool_notifications::CommittedTransaction;
+use aptos_types::transaction::use_case::UseCaseKey;
+use std::{
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap, VecDeque},
+};
+
+pub(crate) struct UseCaseHistory {
+    window_size: usize,
+    num_top_to_track: usize,
+    recent: VecDeque<HashMap<UseCaseKey, usize>>,
+    total: HashMap<UseCaseKey, usize>,
+}
+
+impl UseCaseHistory {
+    pub(crate) fn new(window_size: usize, num_top_to_track: usize) -> Self {
+        Self {
+            window_size,
+            num_top_to_track,
+            recent: VecDeque::with_capacity(window_size + 1),
+            total: HashMap::new(),
+        }
+    }
+
+    fn add_to_recent(&mut self, count_by_usecase: HashMap<UseCaseKey, usize>) {
+        for (use_case, count) in &count_by_usecase {
+            assert!(*count > 0);
+            *self.total.entry(use_case.clone()).or_insert(0) += *count;
+        }
+        self.recent.push_back(count_by_usecase);
+
+        while self.recent.len() > self.window_size {
+            let to_remove = self.recent.pop_front().expect("non-empty after size check");
+            for (use_case, count) in to_remove {
+                assert!(count > 0);
+
+                use std::collections::hash_map::Entry;
+                match self.total.entry(use_case) {
+                    Entry::Occupied(mut o) => {
+                        assert!(*o.get() >= count);
+                        *o.get_mut() -= count;
+                        if *o.get() == 0 {
+                            o.remove_entry();
+                        }
+                    },
+                    Entry::Vacant(e) => {
+                        panic!("Entry present in recent cannot be missing in total {:?}", e)
+                    },
+                }
+            }
+        }
+    }
+
+    pub fn compute_tracking_set(&self) -> HashMap<UseCaseKey, String> {
+        let mut result = HashMap::new();
+        result.insert(UseCaseKey::Platform, "entry_platform".to_string());
+        result.insert(UseCaseKey::Others, "non_entry".to_string());
+
+        let mut max_heap: BinaryHeap<UseCaseByCount> = self
+            .total
+            .iter()
+            .filter(|(use_case, _)| {
+                *use_case != &UseCaseKey::Platform && *use_case != &UseCaseKey::Others
+            })
+            .map(|(use_case, count)| UseCaseByCount {
+                use_case: use_case.clone(),
+                count: *count,
+            })
+            .collect::<BinaryHeap<_>>();
+
+        for i in 0..self.num_top_to_track {
+            if let Some(UseCaseByCount { use_case, count }) = max_heap.pop() {
+                info!(
+                    "Use cases: top {} is {:?} with {} occurences",
```

Comment:
Is this going to be logging too frequently to flood the logs?

Reply:
5 times per block, so 20 per second. 

is that too frequent? I can just merge into single line per block, that's probably fine?

Reply:
That's probably fine but single line per block is better may be?

Reply:
updated